### ISSUE TITLE
-Wunreachable-code-fallthrough doesn't work in templates

### ISSUE THREAD

**Issue by Alcaro** [llvm/llvm-project#58481]
Code:
```c++
int a(int x)
{
    switch (x)
    {
        case 1:
            return 2;
            [[fallthrough]];
        default:
            return 4;
    }
}

template<int b>
int a2(int x)
{
    switch (x)
    {
        case 1:
            return 2;
            [[fallthrough]];
        default:
            return 4;
    }
}

int c()
{
    return a(1) + a2<2>(3) + a2<4>(5);
}
```
Compile with -Wunreachable-code-fallthrough.

Expected result: Two warnings, one for each [[fallthrough]]. Or three, since the template is instantiated twice.

Actual result: One warning, in the untemplated function.

Compiler Explorer: https://godbolt.org/z/cWdoW5WvM

https://github.com/llvm/llvm-project/issues/18993#issuecomment-980927985 says

> (we currentyl don't warn on -Wunreachable-code on any templates because there was a concern that performance would be adversely affected by doing CFG work on lots of uninstantiated templates)

so this may be or have been intentional, but that comment is eight years old, so I'm not sure if it still applies. If it does apply, feel free to wontfix this one.

Found together with @StephanTLavavej on MSVC STL discord

**Comment by dwblaikie** [comment]
Yeah, looks like something got changed that "fixed" #18993 by disabling the warning in templates in general, perhaps.

I think the same problems need to be solved, or maybe the follow-up comment on https://github.com/llvm/llvm-project/issues/18993#issuecomment-980927984 suggests maybe some changes have been made in the right direction? Not sure.

I'd always love to see better modelling for reachability related warnings... (I still think the right solution is to apply it to template patterns rather than instantiations - seeing what the practical perf impact is to builds dealing with lots of uninstantiated templates would be relevant - maybe we have to restrict it to instantiated templates, while still applying it to the pattern not the instantiation)
